lib/Sema/CSSolver.cpp

```diff
@@ -652,6 +652,18 @@ bool ConstraintSystem::tryTypeVariableBindings(
         type = type->reconstituteSugar(/*recursive=*/false);
       }
 
+      // Instead of binding collection types directly let's try to
+      // bind using temporary type variables substituted for element
+      // types, that's going to ensure that subtype relationship is
+      // always preserved.
+      if (isConcreteCollectionType(type)) {
+        auto *BGT = type->getAs<BoundGenericType>();
+        auto UGT = UnboundGenericType::get(BGT->getDecl(), BGT->getParent(),
+                                           BGT->getASTContext());
+
+        type = openUnboundGenericType(UGT, typeVar->getImpl().getLocator());
```

Comment:
It's probably worth calling `reconstituteSugar`, as in the previous code, so we get prettier types.

Reply:
This is effectively dropping all information about the type arguments to `Array` / `Set` / `Dictionary`. I guess that's okay because whatever constraint we inferred this type binding from will still be there to help bind these type variables.

Reply:
Will do re: `reconstituteSugar`, was wondering if we really need it there... Regarding information - yes, that's the intent, to use fresh type variables here so conversion could be done properly.